lib/AST/DiagnosticConsumer.cpp

```diff
@@ -171,22 +185,12 @@ void FileSpecificDiagnosticConsumer::handleDiagnostic(
     consumerSpecificInfo = ConsumerSpecificInfoForSubsequentNotes;
     break;
   }
-  if (!consumerSpecificInfo.hasValue()) {
-    for (auto &subConsumer : SubConsumers) {
-      if (subConsumer.second) {
-        subConsumer.second->handleDiagnostic(SM, Loc, Kind, FormatString,
-                                             FormatArgs, Info);
-      }
-    }
+  if (consumerSpecificInfo.hasValue()) {
+    consumerSpecificInfo.getValue()->subconsumer(*this).handleDiagnostic(SM,Loc, Kind, FormatString, FormatArgs, Info);
     return;
   }
-  if (!consumerSpecificInfo.getValue()->consumer)
-    return; // Suppress non-primary diagnostic in batch mode.
-
-  consumerSpecificInfo.getValue()->consumer->handleDiagnostic(
-      SM, Loc, Kind, FormatString, FormatArgs, Info);
-  consumerSpecificInfo.getValue()->hasAnErrorBeenEmitted |=
-      Kind == DiagnosticKind::Error;
+  for (auto &subconsumer : Subconsumers)
+    subconsumer.handleDiagnostic(SM, Loc, Kind, FormatString, FormatArgs, Info);
```

Comment:
This consolidation came out nicely!

Reply:
Yes, thank you for the original idea! Worked out to move some functions into the substructures.